src/librustc/ty/mod.rs

```diff
@@ -275,6 +275,22 @@ impl ImplOrTraitItemId {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Copy)]
+pub enum Visibility {
+    Public, //< Visible everywhere
+    Restricted(NodeId), //< Visible only in the given module
+    Private, //< Visible nowhere
```

Comment:
I suppose we need to discern between `pub(self)` and private in AST to prohibit things like
```
enum E {
    V { pub(self) variant_field:  u8 }
}
```
(i.e. `SanePrivacyVisitor` should run on post-expansion AST), but in HIR it shouldn't make any difference.